### ISSUE TITLE
Network with too many edges is still visualised

### ISSUE THREAD

**Issue by maehler** [PlantGenIE/PlantGenIE#80]
When loading a network that has more edges than the current limitations, there is a message saying that in order to see the network visualisation, the user has to lower the threshold. This doesn't seem to work since networks that should be too big for visualising are still visualised.

## Steps to reproduce

1. Select *P. tremula* v2.2.
2. Create a gene list with the single gene Potra2n7c15465.
3. Go to exNet
4. Select AspWood
5. Set the threshold as high as it goes (i.e. as far right as possible)
4. Select the single gene in the network
5. Click "Expand network"
6. Click "Select all genes"
7. Click "Create a new GeneList"
8. See the network load with over 12,000 edges, but at the same time you also see the warning message:
> **Connections limit exceeded**
You have >10000 edges and the network display will be too slow, please lower the P-value to see the network.

**Comment by irusri** [comment]
The idea is to show all the edges without filtering by correlation value. However, we warn the user when number of edges are exceeding more than 5000 edges and 10000 edges. So the user is aware that the rendering of the network will be slower and they might need to wait or lower the threshold. 

Now I change the background color of  two warning messages. When the number of edges exceeded 5000, it will give a warning message with green background color. When it exceeded the 10000 limit, it will give a warning message with yellow color background.

If needed I can add the strict cutoff value eg:10000 edges to  show only less 10000 edges maximum limit. Then the user might loose some valuable information. Just let me know what you  think is the best?

**Comment by maehler** [comment]
Yes, but the warning is kind of pointless in that case since it only shows when the network has loaded. Then the damage has already been done. Also, I got confused, because the message says "please lower the P-value to see the network". The way I interpreted that was that the network shouldn't have been shown due to having too many edges.

A couple of things would have to change for this to work better:

- The message has to be shown before the network is drawn. Given that the new slider we were talking about would show the number of edges for the different thresholds, it should be easy to maybe colour that graph something like yellow &rarr; orange &rarr; red when you get into a threshold range that would result in a slow loading of the network. Something like the sketch below?
![edge_graph_warning](https://user-images.githubusercontent.com/2573608/114004451-f6ce6000-985e-11eb-8091-5d3ba352f53c.png)
- The message should be rephrased to read something more like: "You have more than xxx edges in your network, and this might result in long rendering times. Proceed with caution." @natstreet, you're the more frequent user of these tools, so let me know if this would be an appropriate way of phrasing it.

**Comment by maehler** [comment]
Thinking a bit more about the slider, If we have this type of graph, is the p-value display really needed? With the graph we would have an intuitive visualisation of how dense the network would be. We would still have to show the p-value somewhere, but it doesn't have to be together with the slider. I guess this is a question for later though, and not really related to the issue at hand.

**Comment by natstreet** [comment]
I agree with Niklas' last wording suggestion for the message and that the user really needs that information before the action is taken to avoid unknowingly getting stuck with an unresponsive page (if they've been informed and do it anyway then it's on them).

I really like the coloured graph suggestion and think that will work very nicely. We then definitely need the graph to be based on the nodes from there active gene list and not the network globally.

We could just show the slider with labels at either end for something like more stringent (fewer edges) and less stringent (more edges) and then display somewhere else what the selected threshold is. That's probably more how most users will anyway use this rather than thinking purely in terms of 'what p-value is appropriate'.



________________________________

**Comment by irusri** [comment]
great idea! we can update the histogram as a new feature.

I just updated the warning message with "_You have more than 5000/10000 edges in your network, and this might result in long rendering times_.".

**Comment by maehler** [comment]
Still, the message would have to be displayed *before* the network is drawn so that a user could bail out. For me right now, the message is shown *after* the network has been rendered. You could possibly do a request the just replies with the number of edges that would be returned, but I don't know how expensive that would be in the long run. I'm thinking it wouldn't be too bad.

**Comment by irusri** [comment]
with the current setup., I made only one request to the database. However,  we can display the messages right before the network is loaded, then we have to make an additional database query. That is good, if we do not display the network for certain number fo edges. But now we shows the network and we just warn the user about possible bottleneck issues.

**Comment by maehler** [comment]
But then the message is pointless. The bottleneck has already happened, and we're simply explaining for the user why what they just did was slow. In my opinion, this is not very helpful. Then we could just as well remove the message and put a note in the documentation why things might be slow in certain situations.

**Comment by natstreet** [comment]
I agree - what I would suggest is that you make the db request to check how
many nodes/edges will be returned. If that is above a threshold number then
popup and warning that things could be slow and require the user to click
OK to proceed. If the numbers are below the threshold then no popup and
confirmation is needed. It needs to be like this so that the user is warned
before experiencing the wait in my opinion.

/Nat

**Comment by irusri** [comment]
Sure, 
I added the warning messages with OK and Cancel option. When exNet has more than 5000 edges, it will show _You have more than 5000 edges in your network, and this might result in long rendering times. Are you OK with that?_ message. If the user Click OK, it will display the network otherwise it will reload the page.
http://g.recordit.co/l7co8syCoT.gif
